File: staff/src/shared/api/yandexMaps.ts
const API_KEY =
  process.env.NEXT_PUBLIC_YANDEX_MAPS_KEY ?? process.env.NEXT_PUBLIC_YANDEX_MAPS_API_KEY ?? '';

interface SuggestRawTitle {
  text?: string;
}

interface SuggestRawAddress {
  formatted_address?: string;
  full_address?: string;
}

export interface YandexSuggestItem {
  title: string;
  subtitle?: string;
  value: string;
  uri?: string;
}

export const fetchAddressSuggestions = async (
  query: string,
  options: { signal?: AbortSignal } = {}
): Promise<YandexSuggestItem[]> => {
  const trimmed = query.trim();
  if (!API_KEY || !trimmed) {
    return [];
  }

  const params = new URLSearchParams({
    apikey: API_KEY,
    text: trimmed,
    lang: 'ru_RU',
    type: 'geo',
    print_address: '1',
    results: '5',
  });

  const response = await fetch(`https://suggest-maps.yandex.ru/v1/suggest?${params.toString()}`, {
    signal: options.signal,
  });
  if (!response.ok) {
    return [];
  }
  const payload = await response.json();
  const rawResults = Array.isArray(payload?.results) ? (payload.results as unknown[]) : [];
  const suggestions: YandexSuggestItem[] = [];

  for (const item of rawResults) {
    if (!item || typeof item !== 'object') {
      continue;
    }
    const record = item as Record<string, unknown>;
    const rawTitle: SuggestRawTitle | string =
      (record.title as SuggestRawTitle | string | undefined) ?? '';
    const rawSubtitle: SuggestRawTitle | string | undefined = record.subtitle as
      | SuggestRawTitle
      | string
      | undefined;
    const rawAddress: SuggestRawAddress | undefined = record.address as
      | SuggestRawAddress
      | undefined;
    const title = typeof rawTitle === 'string' ? rawTitle : (rawTitle?.text ?? '');
    const subtitle =
      typeof rawSubtitle === 'string' ? rawSubtitle : (rawSubtitle?.text ?? undefined);
    const formatted = rawAddress?.formatted_address ?? rawAddress?.full_address ?? title;
    const value = formatted || title;
    const uri = typeof record.uri === 'string' ? record.uri : undefined;

    if (!value) {
      continue;
    }

    suggestions.push({
      title: title || value,
      subtitle,
      value,
      uri,
    });
  }

  return suggestions;
};

export interface YandexGeocodeResult {
  normalized: string;
  lat: number | null;
  lon: number | null;
  kind: string;
  precision: string;
  uri: string;
}

export const geocodeAddress = async (
  query: string,
  options: { signal?: AbortSignal } = {}
): Promise<YandexGeocodeResult | null> => {
  const trimmed = query.trim();
  if (!API_KEY || !trimmed) {
    return null;
  }

  const params = new URLSearchParams({
    apikey: API_KEY,
    geocode: trimmed,
    format: 'json',
    lang: 'ru_RU',
  });

  const response = await fetch(`https://geocode-maps.yandex.ru/v1/?${params.toString()}`, {
    signal: options.signal,
  });
  if (!response.ok) {
    return null;
  }

  const payload = await response.json();
  const collection = payload?.response?.GeoObjectCollection ?? {};
  const rawMembers = Array.isArray(collection?.featureMember)
    ? (collection.featureMember as unknown[])
    : [];
  if (rawMembers.length === 0) {
    return null;
  }

  const firstMember = rawMembers[0];
  if (!firstMember || typeof firstMember !== 'object') {
    return null;
  }

  const geoObjectRecord = firstMember as Record<string, unknown>;
  const geoObject = (geoObjectRecord.GeoObject as Record<string, unknown> | undefined) ?? {};
  const metaDataProperty = geoObject.metaDataProperty as Record<string, unknown> | undefined;
  const geocoderMeta = metaDataProperty?.GeocoderMetaData as Record<string, unknown> | undefined;
  const address = geocoderMeta?.Address as Record<string, unknown> | undefined;
  const posSource = geoObject.Point as Record<string, unknown> | undefined;
  const pos = typeof posSource?.pos === 'string' ? posSource.pos : '';
  const [lonStr, latStr] = pos.split(' ');
  const lat = latStr ? Number(latStr) : null;
  const lon = lonStr ? Number(lonStr) : null;
  const normalized =
    (address?.formatted as string | undefined) ??
    (geocoderMeta?.text as string | undefined) ??
    trimmed;
  const kind = typeof geocoderMeta?.kind === 'string' ? (geocoderMeta.kind as string) : '';
  const precision =
    typeof geocoderMeta?.precision === 'string' ? (geocoderMeta.precision as string) : '';
  const uri = typeof geoObject?.uri === 'string' ? (geoObject.uri as string) : '';

  return {
    normalized,
    lat: Number.isFinite(lat) ? lat : null,
    lon: Number.isFinite(lon) ? lon : null,
    kind,
    precision,
    uri,
  };
};

File: backend/applications/orders/services/yandex_maps.py
from __future__ import annotations

import os
from decimal import ROUND_HALF_UP, Decimal
from math import atan2, cos, radians, sin, sqrt
from typing import Any

import requests

YANDEX_MAPS_KEY = os.environ.get('YANDEX_MAPS_SECRET', '')


class YandexMapsError(RuntimeError):
    """Raised when Yandex Maps API responds with an error."""


def _get_api_key() -> str:
    key = YANDEX_MAPS_KEY or os.environ.get('YANDEX_MAPS_SECRET', '')
    if not key:
        raise YandexMapsError('YANDEX_MAPS_KEY is not configured')
    return key


def geocode_address(
    query: str, *, lang: str = 'ru_RU', timeout: float = 7.0
) -> dict[str, Any] | None:
    """Resolve address details using Yandex Geocoder API.

    Returns a dictionary with normalized address data or ``None`` when the
    address could not be resolved.
    """

    query = (query or '').strip()
    if not query:
        return None

    params = {
        'apikey': _get_api_key(),
        'geocode': query,
        'lang': lang,
        'format': 'json',
    }
    headers = {'Referer': os.environ.get('YandexReferer', '')}  # апи яндекс карт - величие
    response = requests.get(
        'https://geocode-maps.yandex.ru/v1/', params=params, headers=headers, timeout=timeout
    )
    response.raise_for_status()

    payload = response.json()
    collection = payload.get('response', {}).get('GeoObjectCollection', {})
    members = collection.get('featureMember', [])
    if not members:
        return None

    geo_object = members[0]['GeoObject']
    meta = geo_object.get('metaDataProperty', {}).get('GeocoderMetaData', {})
    position = geo_object.get('Point', {}).get('pos', '')
    try:
        lon_str, lat_str = position.split()
        lon = float(lon_str)
        lat = float(lat_str)
    except (ValueError, AttributeError):
        lon = lat = None

    normalized = meta.get('Address', {}).get('formatted') or meta.get('text')
    return {
        'normalized': normalized or '',
        'lat': lat,
        'lon': lon,
        'kind': meta.get('kind', ''),
        'precision': meta.get('precision', ''),
        'uri': geo_object.get('uri', ''),
    }


def _request_router_distance(
    origin: tuple[float, float], destination: tuple[float, float], timeout: float
) -> Decimal:
    params = {
        'apikey': _get_api_key(),
        'mode': 'driving',
        'waypoints': f'{origin[1]},{origin[0]}|{destination[1]},{destination[0]}',
    }
    response = requests.get(
        'https://api.routing.yandex.net/v2/route', params=params, timeout=timeout
    )
    response.raise_for_status()
    payload = response.json()
    routes = payload.get('routes') or []
    if not routes:
        raise YandexMapsError('Маршрут не найден в ответе Яндекс.Карт.')
    try:
        distance_meters = routes[0]['legs'][0]['distance']['value']
    except (KeyError, IndexError, TypeError) as exc:
        raise YandexMapsError('Некорректный ответ от API маршрутизации Яндекс.') from exc
    return (Decimal(distance_meters) / Decimal('1000')).quantize(
        Decimal('0.01'), rounding=ROUND_HALF_UP
    )


def _geodesic_distance(origin: tuple[float, float], destination: tuple[float, float]) -> Decimal:
    radius_km = 6371.0
    lat1, lon1 = origin
    lat2, lon2 = destination
    dlat = radians(lat2 - lat1)
    dlon = radians(lon2 - lon1)
    a = sin(dlat / 2) ** 2 + cos(radians(lat1)) * cos(radians(lat2)) * sin(dlon / 2) ** 2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    return Decimal(radius_km * c).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)


def calculate_route_distance_km(
    origin_address: str, destination_address: str, *, timeout: float = 7.0
) -> Decimal:
    """Calculate distance between two addresses using Yandex Maps data."""
    origin = geocode_address(origin_address)
    destination = geocode_address(destination_address)
    if not origin or not destination:
        raise YandexMapsError('Не удалось получить координаты для расчёта расстояния.')
    if None in (
        origin.get('lat'),
        origin.get('lon'),
        destination.get('lat'),
        destination.get('lon'),
    ):
        raise YandexMapsError('Яндекс.Карты не вернули координаты адреса.')
    origin_coords = (float(origin['lat']), float(origin['lon']))
    destination_coords = (float(destination['lat']), float(destination['lon']))
    try:
        return _request_router_distance(origin_coords, destination_coords, timeout)
    except (requests.RequestException, YandexMapsError):  # pragma: no cover - network fallback
        return _geodesic_distance(origin_coords, destination_coords)
